### PR TITLE
[ADP-3256] Create conway era based local-cluster

### PR DESCRIPTION
**Pull request by paolino**
This PR add the support for a Conway era based local-cluster and create a Conway integration tests suite.

- [x] Remove Vasil HF dead code
- [x] Restructure Cluster module
- [x] Add ConwayHardFork constructor
- [x] Add logging the era the cluster is running into (as per config)
- [x] Restructure integration tests executable into sub-modules
- [x] Add ClusterM monad to hold the Config as reader context
- [x] Removed Tagged in favor of a newtype FileOf
- [x] Removed Tagged "output" and Tagged "pool" as computable from Config
- [x] Create Conway integration test executable, exposed in the nix shell
- [x] Add tests that do not require cluster to work in the Conway suite
- [x] Add a no-op test that require the Conway cluster
- [x] Removed MIR use to allow the Conway cluster to start (MIR not supported anymore)   
- [x] Create Babbage integration test executable, exposed in the nix shell
- [x] Add just commands to run both integration test suite    
- [x] Add both test suites to the buildkite pipeline

ADP-3256

### CODE REVIEW

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/ClusterEra.hs [hunk 1]
```diff
@@ -42,23 +42,27 @@ data ClusterEra
...
-    fmap withDefault . traverse getEra =<< lookupEnvNonEmpty var
+clusterEraFromEnv = do
+    mera <- lookupEnvNonEmpty var
+    case mera of
+        Nothing -> pure BabbageHardFork
+        Just era -> getEra era
   where
     var = "LOCAL_CLUSTER_ERA"
+    err :: [Char] -> IO a
+    err era = die $ var ++ ": " ++ era ++ " era is not supported"
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/ClusterEra.hs]: 👍 Thanks

Ideally we should drop the pre-Babbage constructors from `ClusterEra` as well, not just tweak `clusterEraFromEnv`.

File: lib/wallet/test/integration/shelley-integration-test.hs [hunk 2]
```diff
@@ -385,7 +385,7 @@ specWithServer testnetMagic testDir (tr, tracers) = aroundAll withContext
                     }
 
                 era <- clusterEraFromEnv
-
+                traceWith tr $ MsgCluster $ Cluster.MsgHardFork era
```
Comment (Anviking) [inline lib/wallet/test/integration/shelley-integration-test.hs]: I think this belongs in `withCluster` itself, not here. Could also potentially be merged with `MsgStartingCluster`

File: lib/local-cluster/lib/Cardano/Wallet/LocalCluster.hs [hunk 3]
```diff
@@ -217,10 +217,11 @@ main = withUtf8 $ do
                 , cfgClusterConfigs = clusterConfigsDir
                 , cfgTestnetMagic = Cluster.TestnetMagic 42
                 , cfgShelleyGenesisMods = [ over #sgSlotLength \_ -> 0.2 ]
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/LocalCluster.hs]: 👍 could confuse us that we define this value twice, but I hope we manage

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/CardanoCLI.hs [hunk 4]
```diff
@@ -61,49 +75,49 @@ import qualified Data.ByteString.Lazy.Char8 as BL8
...
-                -> ProcessConfig stdin stdout stderr
+    let cliEnv
+            :: ProcessConfig stdin stdout stderr
+            -> ProcessConfig stdin stdout stderr
         cliEnv = maybe setEnvInherit (setEnv . mkEnv) conn
-    pure $ cliEnv $ proc "cardano-cli" args
+    pure
+        $ cliEnv
+        $ proc "cardano-cli"
+        --  $ clusterEraToString cfgLastHardFork :
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/CardanoCLI.hs]: Commented line

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/CardanoCLI.hs [hunk 5]
```diff
@@ -61,49 +75,49 @@ import qualified Data.ByteString.Lazy.Char8 as BL8
...
     -- ^ command-line arguments
-    -> IO (ProcessConfig () () ())
-cliConfigBase tr conn args = do
-    traceWith tr (MsgCLI args)
-    env <- getEnvironment
+    -> ClusterM (ProcessConfig () () ())
+cliConfigBase conn args = do
+    traceClusterLog (MsgCLI args)
+    env <- liftIO getEnvironment
+    -- Config{..} <- ask
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/CardanoCLI.hs]: Commented line

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/PoolMetadataServer.hs [hunk 6]
```diff
@@ -58,13 +58,13 @@ withPoolMetadataServer
...
 withPoolMetadataServer action = do
-    config@Config{..} <- ask
+    UnliftClusterM withConfig Config{..} <- askUnliftClusterM
     let metadir = untag cfgClusterDir </> "pool-metadata"
     liftIO $ do
         createDirectoryIfMissing False metadir
         withStaticServer metadir $ \baseURL -> do
             let _urlFromPoolIndex i = baseURL </> metadataFileName i
-            runClusterM config $ action PoolMetadataServer
+            withConfig $ action PoolMetadataServer
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/PoolMetadataServer.hs]: Is `ClusterM` worth it compared to taking a `Config` argument? 🤔 Any particular reason for introducing it?

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Config.hs [hunk 7]
```diff
@@ -62,3 +64,7 @@ data Config = Config
     -- ^ Shelley genesis modifications to apply.
     , cfgTracer :: Tracer IO ClusterLog
     }
+
+newtype NodeSegment = NodeSegment FilePath
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Config.hs]: - `NodePathSegment`?
- `NodeDirName`?

File: lib/wallet/test/integration/shelley-integration-test.hs [hunk 8]
```diff
@@ -108,7 +114,7 @@ main = withTestsSetup $ \testDir (tr, tracers) -> do
             -- same .tix file simultaneously, causing errors.
             --
             -- Because of this, don't run the CLI tests in parallel in CI.
-            parallelIf (not nix) $ describe "CLI Specifications" $ do
+            parallelIf True $ describe "CLI Specifications" $ do
```
Comment (Anviking) [inline lib/wallet/test/integration/shelley-integration-test.hs]: 👍 but let's clean up this `parallelIf` and comment then

File: lib/wallet/cardano-wallet.cabal [hunk 9]
```diff
@@ -806,16 +806,26 @@ common integration-common
...
   type:               exitcode-stdio-1.0
   main-is:            babbage-integration-tests.hs
   build-tool-depends: cardano-wallet:cardano-wallet
 
+test-suite conway-integration
+  import:             integration-common
+  type:               exitcode-stdio-1.0
+  main-is:            conway-integration-tests.hs
+  build-tool-depends: cardano-wallet:cardano-wallet
+
```
Comment (Anviking) [inline lib/wallet/cardano-wallet.cabal]: Traditionally we have had a single test suite, at times with some era-dependent expectations. Why not do it this way? E.g. by marking all tests `pendingOnConway` for now

File: flake.nix [hunk 10]
```diff
@@ -250,31 +250,7 @@
...
-                            # Uncomment and update the LOCAL_CLUSTER_ERA when we're close to a
-                            # hard-fork and would like to run integration tests on two eras in CI
-                            # again.
-                            # cardano-wallet.integration-prev-era = integrationPrevEraCheck;
-                          }
-                        else
-                          {}
-                    )
-                );
+                checks = lib.removeRecurse (collectChecks isProjectPackage coveredProject.hsPkgs);
```
Comment (Anviking) [inline flake.nix:252]: ✔️

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Cluster.hs [hunk 11]
```diff
@@ -136,6 +137,9 @@ import Control.Monad.Reader
     ( MonadIO (..)
     )
 
+import Cardano.Wallet.Launch.Cluster.ClusterEra
+    ( ClusterEra (BabbageHardFork)
+    )
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Cluster.hs]: Wrong section

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Cluster.hs [hunk 12]
```diff
@@ -248,7 +252,8 @@ withCluster config@Config{..} faucetFunds onClusterStart = runClusterM config
         let RunningNode conn _ _ = runningNode
 
         -- Needs to happen in the first 20% of the epoch, so we run this first.
-        moveInstantaneousRewardsTo conn mirCredentials
+        when (cfgLastHardFork == BabbageHardFork)
+            $ moveInstantaneousRewardsTo conn mirCredentials
```
Comment (Anviking) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Cluster.hs:257]: 👍 for now, let's see what we do later

File: lib/wallet/cardano-wallet.cabal (hunk 9 above)
Comment (Anviking) [reply lib/wallet/cardano-wallet.cabal]: One current oddity I think is `conway-integration` and `babbage-integration` both having `BabbageEra` as default era...

**Review by Anviking — COMMENTED** [review]
Thanks, largely LGTM, <s>though it could have been easier to review if split into refactoring and functionality-extending PRs</s> just saying I trust you on a bunch of the details.

My main reservation would be why not to keep Babbage and Conway integration tests as one.

Happy to discuss details...

File: lib/local-cluster/lib/Cardano/Wallet/LocalCluster.hs (hunk 3 above)
Comment (Anviking) [reply lib/local-cluster/lib/Cardano/Wallet/LocalCluster.hs]: Ok, I see it's like this on master already

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/ClusterEra.hs (hunk 1 above)
Comment (paolino) [reply lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/ClusterEra.hs]: They are used somewhere else ..

File: lib/wallet/test/integration/shelley-integration-test.hs (hunk 2 above)
Comment (paolino) [reply lib/wallet/test/integration/shelley-integration-test.hs]: I agree, I moved it here to spot the message next to EKG

File: lib/wallet/cardano-wallet.cabal (hunk 9 above)
Comment (paolino) [reply lib/wallet/cardano-wallet.cabal]: Yes we need to discuss this choice.

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/PoolMetadataServer.hs (hunk 6 above)
Comment (paolino) [reply lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/PoolMetadataServer.hs]: Yes , the Config is needed in code "leaves" so passing it around is a bit inconvenient in the local-cluster code, so I used everywhere in the end

**Review by Anviking — APPROVED** [review]
LGTM, thanks! 🙌

**Pull request merged**